src/cache_cf.cc

```diff
@@ -1014,87 +1017,133 @@ parse_obsolete(const char *name)
-{
-    time_msec_t u = parseTimeUnits(units, allowMsec);
-    if (u == 0) {
-        self_destruct();
-        return;
-    }
+/// Assigns 'ns' the number of nanoseconds corresponding to 'unitName'.
+/// \param MinimalUnit is a chrono duration type specifying the minimal
+/// allowed time unit.
+/// \returns true if unitName is correct and its time unit is not less
+/// than MinimalUnit.
+template <class MinimalUnit>
+bool
+parseSingleTimeUnit(const char *unitName, std::chrono::nanoseconds &ns)
+{
+    assert(unitName);
+    if (!strncasecmp(unitName, T_NANOSECOND_STR, strlen(T_NANOSECOND_STR)))
+        ns = std::chrono::nanoseconds(1);
+    else if (!strncasecmp(unitName, T_MICROSECOND_STR, strlen(T_MICROSECOND_STR)))
+        ns = std::chrono::microseconds(1);
+    else if (!strncasecmp(unitName, T_MILLISECOND_STR, strlen(T_MILLISECOND_STR)))
+        ns = std::chrono::milliseconds(1);
+    else if (!strncasecmp(unitName, T_SECOND_STR, strlen(T_SECOND_STR)))
+        ns = std::chrono::seconds(1);
+    else if (!strncasecmp(unitName, T_MINUTE_STR, strlen(T_MINUTE_STR)))
+        ns = std::chrono::minutes(1);
+    else if (!strncasecmp(unitName, T_HOUR_STR, strlen(T_HOUR_STR)))
+        ns = std::chrono::hours(1);
+    else if (!strncasecmp(unitName, T_DAY_STR, strlen(T_DAY_STR)))
+        ns = std::chrono::hours(24);
+    else if (!strncasecmp(unitName, T_WEEK_STR, strlen(T_WEEK_STR)))
+        ns = std::chrono::hours(24 * 7);
+    else if (!strncasecmp(unitName, T_FORTNIGHT_STR, strlen(T_FORTNIGHT_STR)))
+        ns = std::chrono::hours(24 * 14);
+    else if (!strncasecmp(unitName, T_MONTH_STR, strlen(T_MONTH_STR)))
+        ns = std::chrono::hours(24 * 30);
+    else if (!strncasecmp(unitName, T_YEAR_STR, strlen(T_YEAR_STR)))
+        ns = std::chrono::hours(static_cast<std::chrono::hours::rep>(HoursPerYear));
+    else if (!strncasecmp(unitName, T_DECADE_STR, strlen(T_DECADE_STR)))
+        ns = std::chrono::hours(static_cast<std::chrono::hours::rep>(HoursPerYear * 10));
+    else
+        return false;
 
-    char *token = ConfigParser::NextToken();;
-    if (!token) {
+    if (ns < MinimalUnit(1)) {
+        debugs(3, DBG_CRITICAL, "time unit '" << unitName << "' is too small to be used in this context");
         self_destruct();
-        return;
+        return false;
     }
 
-    double d = xatof(token);
-
-    time_msec_t m = u; /* default to 'units' if none specified */
+    return true;
+}
 
-    if (d) {
-        if ((token = ConfigParser::PeekAtToken()) && (m = parseTimeUnits(token, allowMsec))) {
-            (void)ConfigParser::NextToken();
+/// Wraps parseSingleTimeUnit() and parses defaultUnitName if it failed to parse unitName.
+/// \returns true iff unitName was successfully parsed.
+template <class MinimalUnit>
+bool
+parseTimeUnits(const char *unitName, const char *defaultUnitName, std::chrono::nanoseconds &ns)
+{
+    if (unitName && parseSingleTimeUnit<MinimalUnit>(unitName, ns))
```

Comment:
This "and" logic combined with the code immediately below overwrites bad unit name if there is a default unit name. Yet, there is also code further below that declares `unknown time unit` a FATAL error. Can we unify/simplify this using the `if a unit name is given, it must be valid` principle and then dying on invalid unit names?

I suspect the answer is "not easily because we want to handle such-and-such configurations". If that is the case, please give an example (or two) of squid.conf directives that necessitate such an awkward interface. I wonder whether we can just stop supporting them instead of bending over backwards to backward compatibility sake.

Reply:
> Can we unify/simplify this using the if a unit name is given, it must be valid principle

I think we can. 

> I suspect the answer is "not easily because we want to handle such-and-such configurations".

Squid has a 'url_rewrite_timeout' parameter, requiring special parsing because its 'time-units' option may be followed by other parameters. This is the only parameter (AFAIK) allowing default timeunit (seconds), which is used when there is no explicit timeunit. If the explicit timeunit is invalid, the default unit is used either (overwriting this invalid one as you noted). However, in this case, the token is not consumed, and the parsing aborts because it cannot parse the next  'url_rewrite_timeout' option. In other words, Squid complains about the 'unsupported option' rather than about the 'unknown time unit'. I did not change this behavior, but probably we can follow your suggestion and complain about an 'unknown time unit' instead (Squid terminates anyway).

Reply:
AFAICT from the [documented](http://www.squid-cache.org/Doc/config/url_rewrite_timeout/) url_rewrite_timeout syntax, the time units are _required_. If unpatched Squid (v4 and master) dies when time units are missing (regardless of the exact death place in code and the accompanying error message), then we may assume that the time unit is required as far as admins and their old configurations are concerned, even if there is (broken) code that implies otherwise.

If proceeding down that path simplifies this code/APIs, then we should simplify this code/APIs accordingly (and briefly justify dropping that broken code in the PR description).

From your description, it is not clear to me whether time units are indeed required (from admin and old configurations points of view). Is it possible to omit the unit and successfully configure Squid with `url_rewrite_timeout 5 on_timeout=fail`, for example?

If time unit is effectively optional for url_rewrite_timeout, then we may still want to simplify this code/API by adding url_rewrite_timeout-only code that

1. peeks at the second token,
2. determines whether it is a time unit (by comparing that token with all known unit names), and
3. branches into "explicit time unit required" and "implicit time unit used" code, each branch using simple/common APIs to handle its now-known syntax.

Reply:
> Is it possible to omit the unit and successfully configure Squid with url_rewrite_timeout 5 on_timeout=fail, for example?

It is/was possible, and a default timeunit (second in this case) is provided then. Also a level-0 warning is reported: `WARNING: No units on [...], assuming [...]`.

Reply:
Since that `WARNING` was present in unpatched Squid, we have even more ammunition to discontinue that (wrongheaded) support for guessing units. However, to minimize time wasted on the official review, I would still keep that support _if_ it is easy to write the necessary code in the caller, with a deprecation warning added.

Reply:
Done at ba0b138.